### DISCUSSION TITLE
[Reflection] How do I convert my MethodInfo definition to a MethodInfo where the declaring type is a generic type instance?

### DISCUSSION THREAD

**Discussion by slxdy** [dotnet/runtime#99591] [Q&A] [answered]
Right now, I have this code
```cs
class Program
{
    static void Main()
    {
        var method = typeof(GenericT<>).GetMethod("PrintName", BindingFlags.Static | BindingFlags.NonPublic);
        var genInstanceMet = // Convert 'method' from 'GenericT<>::PrintName' to 'GenericT<string>::PrintName'
        genInstanceMet.Invoke(null, null);
    }
}

public class GenericT<T>
{
    private static void PrintName()
    {
        Console.WriteLine(typeof(T).FullName);
    }
}
```

How do I convert the `method` from a definition to a reference where the declaring type is a generic type instance?

Edit: I mean without modifying the line where I define the `method` variable

**Comment by KeterSCP** [comment]
Not sure if this is 100% correct, but you can just make a generic type from method's info declaring type:

```csharp
class Program
{
    static void Main()
    {
        var method = typeof(GenericT<>).GetMethod("PrintName", BindingFlags.Static | BindingFlags.NonPublic);
        var genInstanceMet = method.DeclaringType.MakeGenericType(typeof(string)).GetMethod(method.Name, BindingFlags.Static | BindingFlags.NonPublic);
        genInstanceMet.Invoke(null, null); // prints System.String
    }
}

public class GenericT<T>
{
    private static void PrintName()
    {
        Console.WriteLine(typeof(T).FullName);
    }
}
```

**Reply by MichalStrehovsky** [reply]
This will work if there is no other method named PrintName. Otherwise you need to disambiguate signatures and that may quickly become complicated:

```csharp
public class GenericT<T>
{
    private static void PrintName(string s)
    {
        Console.WriteLine(typeof(T).FullName);
    }

    private static void PrintName(T s)
    {
        Console.WriteLine(typeof(T).FullName);
    }
}
```

**Comment by MichalStrehovsky** [comment, marked as answer]
```csharp
var genInstanceMet = (MethodInfo)typeof(GenericT<string>).GetMemberWithSameMetadataDefinitionAs(method);
```

**Reply by slxdy** [reply]
Perfect, thanks!
However, I have one more question, because my end goal is a bit more complicated.
How would I do this for a dynamic method where the dynamic type hasn't been created yet?
I tried using this solution, but it just throws a `System.NotSupportedException: 'Specified method is not supported.'`
How would I accomplish that?

**Reply by MichalStrehovsky** [reply]
Try if `MemberInfo.HasSameMetadataDefinitionAs` works. If it works, you'd need to do a foreach over all methods on the instantiated type and find the one that returns true.

If it doesn't work, I think that's it. There's MemberInfo.MetadataToken that you could also compare, but that probably also doesn't work for dynamic methods.

**Reply by slxdy** [reply]
About that....
![image](https://github.com/dotnet/runtime/assets/61495410/4830cb5c-80a2-46f5-92f7-38839218f3ed)

Should I open a new issue, considering that this is a limitation to Reflection.Emit? There is no existing way to get a dynamic method reference with a declaring type that has generic args defined.

**Reply by MichalStrehovsky** [reply]
> Should I open a new issue

Worth a shot.

**Reply by slxdy** [reply]
No wait, there seems to be an implementation for this at https://github.com/dotnet/runtime/blob/8790a2ffd49e9d3b9bf1621525718c753c28d519/src/libraries/System.Private.CoreLib/src/System/Reflection/Emit/MethodOnTypeBuilderInstantiation.cs

**Reply by slxdy** [reply]
Alright, I found the answer
`TypeBuilder::GetMethod`

I wish the static TypeBuilder functions were more exposed.
